---
title: "Tutorial for GWAS with SLOPE"
author: "Piotr Sobczyk"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial for GWAS with SLOPE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Tutorial on performing GWAS with SLOPE

```{r, results='hide', message=FALSE}
library(cps)
pheFile <- system.file("extdata", "plinkPhenotypeExample.phe", package = "cps")
snpsFile <- system.file("extdata", "plinkDataExample.tped", package = "cps")
```

```{r}
phe <- readPhenotype(filename = "../phenotype.csv", sep=";")
```

Now we can read in snp data. We assume that snps were previously exported from PLINK
as **transposed** matrix. To do that use PLINK command

**plink --file data --recode12 --transposed --out temp}**

where data is you name of .ped file. Depending on data size this make take some time.
As data is very large, snps are filtered with their marginal test p-value.
All snps which p-values are large than threshold *pValMax* will be truncated.

```{r, warning=FALSE}
screening <- readSNPs("../plink.tped", phe$y, pValMax = 0.05, chunk.size = 2e4)
```

User can check what is result of screening

```{r}
summary(screening)
```

Next step is clumping. Highly correlated snps will be clustered. When *rho*
decreases so does number of clumps; however, clumps are getting larger.

```{r}
clumping <- clumpProcedure2(screening, rho = 0.3)
```

What is the result of clumping?

```{r}
summary(clumping)
```

We can also plot our results

```{r}
plot(clumping)
```

If we are interested in specific genome we can "zoom it"

```{r}
plot(clumping, 2)
```


Last step of analysis is using SLOPE

```{r}
slope.result <- genSLOPE(clumping, fdr=0.1)
```

As before one can plot and summarize results

```{r}
summary(slope.result)
plot(slope.result)
```

#### Detailed description of parameters

There are three numerical parameters that influence result

* *pValMax* is the threshold p-value for marginal test. When data is loaded to R,
initial screening of snps is performed. For every snp, test for coefficient in simple linear regression model *lm(phenotype~snp)* is performed. All snps with p-value larger than pValMax are discarded. Setting this parameter to too large will significantly increase number of snps on which clumping procedure will be performed. This may cause two technical threats
    * Computer might run out of RAM memory
    * Clumping procedure might take a lot of time
* *rho* is threshold for correlation between snps in clumping procedure. For given snp, every that is correlated with it at least at level *rho* will be clumped together. Setting this parameter too high (say 0.7) will cause SLOPE to work on highly correlated snps which might affect its accuracy
* *fdr* false discover rate for SLOPE procedure. The higher FDR is, more variables will be accepted to the model. Contrary, small fdr yields more conservative models

We set verbose to FALSE to suppress progress bar. 
